tools/loader/ImageClassifier.cpp

```diff
@@ -153,68 +120,46 @@ llvm::cl::opt<bool> useImagenetNormalization(
-  auto range = normModeToRange(imageNormMode);
   unsigned numImages = filenames.size();
 
-  // Get first image's dimensions and check if grayscale or color.
-  size_t imgHeight, imgWidth;
-  bool isGray;
-  std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filenames[0].c_str());
-  const size_t numChannels = isGray ? 1 : 3;
-
-  // N x C x H x W
-  inputImageData->reset(ElemKind::FloatTy,
-                        {numImages, numChannels, imgHeight, imgWidth});
+  Tensor initial =
+      readPngImageAndPreprocess(filenames[0], imageNormMode, imageChannelOrder,
+                                imageLayout, useImagenetNormalization);
+  auto IH = initial.getHandle();
+  inputImageData->reset(
+      ElemKind::FloatTy,
+      {numImages, initial.dims()[0], initial.dims()[1], initial.dims()[2]});
   auto IIDH = inputImageData->getHandle<>();
-
-  // We iterate over all the png files, reading them all into our inputImageData
-  // tensor for processing
-  for (unsigned n = 0; n < filenames.size(); n++) {
-    Tensor localCopy;
-    // PNG images are loaded as NHWC & RGB
-    bool loadSuccess = !readPngImage(&localCopy, filenames[n].c_str(), range,
-                                     useImagenetNormalization);
-    GLOW_ASSERT(loadSuccess && "Error reading input image.");
-    auto imageH = localCopy.getHandle<>();
-
-    auto dims = localCopy.dims();
-    assert((dims[0] == imgHeight && dims[1] == imgWidth) &&
-           "All images must have the same Height and Width");
-    assert(dims[2] == numChannels &&
-           "All images must have the same number of channels");
-    assert((imageChannelOrder == ImageChannelOrder::BGR ||
-            imageChannelOrder == ImageChannelOrder::RGB) &&
-           "Invalid image format");
-
-    // Convert to NCHW with the requested channel ordering.
-    for (unsigned z = 0; z < numChannels; z++) {
-      for (unsigned y = 0; y < dims[1]; y++) {
-        for (unsigned x = 0; x < dims[0]; x++) {
-          if (imageChannelOrder == ImageChannelOrder::BGR) {
-            IIDH.at({n, numChannels - 1 - z, x, y}) = (imageH.at({x, y, z}));
-          } else { // RGB
-            IIDH.at({n, z, x, y}) = (imageH.at({x, y, z}));
-          }
-        }
-      }
-    }
-  }
-
-  // For ONNX graphs with input in NHWC layout, we transpose the image data.
-  switch (imageLayout) {
-  case ImageLayout::NCHW:
-    break;
-  case ImageLayout::NHWC:
-    Tensor dataNHWC;
-    inputImageData->transpose(&dataNHWC, NCHW2NHWC);
-    *inputImageData = std::move(dataNHWC);
-    break;
+  insertImage(IIDH, IH, 0);
+
+  for (size_t n = 1; n < filenames.size(); n++) {
+    Tensor localCopy = readPngImageAndPreprocess(filenames[n], imageNormMode,
+                                                 imageChannelOrder, imageLayout,
+                                                 useImagenetNormalization);
+    assert(std::equal(localCopy.dims().begin(), localCopy.dims().end(),
```

Comment:
nit: Can you do `localCopy.dims() == initial.dims()` instead?

Reply:
Hm, now that I'm using getPngInfo, I no longer have `initial`, so I'm using `std::equal` with `batch.dims().begin() + 1`.  What do you think of that?  It's kind of weird, but I don't have a better idea :-/

Reply:
You could create a const array above with the expected shape `{imgHeight, imgWidth, numChannels}` and then compare here. NBD either way though.